rust/xaynet-analytics/src/controller.rs

```diff
@@ -148,3 +176,378 @@ impl AnalyticsController {
+        let timestamp = DateTime::parse_from_rfc3339("2021-01-01T01:01:00+00:00")
+            .unwrap()
+            .with_timezone(&Utc);
+        assert!(controller
+            .save_analytics_event(name, event_type, timestamp, None)
+            .is_ok());
+
+        let analytics_event = AnalyticsEvent::new(name, event_type, timestamp, None);
+        let all_analytics_events =
+            AnalyticsEvent::get_all(controller.db(), CollectionNames::ANALYTICS_EVENTS);
+        assert_eq!(
+            all_analytics_events.unwrap().first(),
+            Some(&analytics_event)
+        );
+
+        cleanup(controller, test_name);
+    }
+
+    #[test]
+    fn test_save_analytics_event_with_screen_route() {
+        let test_name = "test_save_analytics_event_with_screen_route";
+        let controller = get_controller(test_name, None);
+        let name = "test";
+        let event_type = AnalyticsEventType::ScreenEnter;
+        let timestamp = DateTime::parse_from_rfc3339("2021-01-01T01:01:00+00:00")
+            .unwrap()
+            .with_timezone(&Utc);
+        let screen_route_name = "route";
+
+        assert!(controller
+            .save_analytics_event(name, event_type, timestamp, Some(screen_route_name))
+            .is_ok());
+
+        let screen_route = ScreenRoute::new(screen_route_name, timestamp);
+        let analytics_event = AnalyticsEvent::new(name, event_type, timestamp, Some(screen_route));
+        let all_analytics_events =
+            AnalyticsEvent::get_all(controller.db(), CollectionNames::ANALYTICS_EVENTS);
+        assert_eq!(
+            all_analytics_events.unwrap().first(),
+            Some(&analytics_event)
+        );
+
+        cleanup(controller, test_name);
+    }
+
+    #[test]
+    fn test_change_connectivity_status() {
+        let test_name = "test_change_connectivity_status";
+        let mut controller = get_controller(test_name, None);
+        assert_eq!(controller.is_connected_to_wifi, true);
+
+        controller.change_connectivity_status();
+        assert_eq!(controller.is_connected_to_wifi, false);
+
+        cleanup(controller, test_name);
+    }
+
+    #[test]
+    fn test_change_state_of_charge() {
+        let test_name = "test_change_state_of_charge";
+        let mut controller = get_controller(test_name, None);
+        assert_eq!(controller.is_charging, true);
+
+        controller.change_state_of_charge();
+        assert_eq!(controller.is_charging, false);
+
+        cleanup(controller, test_name);
+    }
+
+    #[test]
+    fn test_validate_send_data_frequency_when_none() {
+        assert_eq!(
+            AnalyticsController::validate_send_data_frequency(None).unwrap(),
+            Duration::hours(AnalyticsController::MAX_SEND_DATA_FREQUENCY_HOURS as i64)
+        )
+    }
+
+    #[test]
+    fn test_validate_send_data_frequency_when_more_than_24() {
+        assert!(AnalyticsController::validate_send_data_frequency(Some(25)).is_err());
```

Comment:
Good point. Added a test. it becomes `Duration::hours(0)`, so we would always send as long as the device is connected to wifi and charging. Not sure if that's what we want, but I guess 0 is still a valid input?
On the Flutter/Dart side there will still need to be a background process that cannot run more often than each 15 minutes i think (because of rules of the OS), and we cannot bury it only on the Rust side. @Robert-Steiner knows more about this. 

See [here](https://github.com/xaynetwork/xaynet-sdk-dart-testing-ground/blob/master/example/lib/main.dart) for some reference (the BackgroundFetch stuff)